projects/centre_console/src/drive_fsm.c

```diff
@@ -0,0 +1,294 @@
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_DRIVE, state_set_ebrake);
+}
+
+FSM_STATE_TRANSITION(state_reverse) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_PARKING, state_set_relays);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_DRIVE, state_switch_direction);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_NEUTRAL, state_set_relays);
+}
+
+FSM_STATE_TRANSITION(state_switch_direction) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_REVERSE, state_reverse);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_DRIVE, state_drive);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+FSM_STATE_TRANSITION(state_fault) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_OPENED_DESTINATION_NEUTRAL_PARKING,
+                     state_neutral);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+static bool prv_destination_guard(const Fsm *fsm, const Event *e, void *context) {
+  DriveFsmStorage *storage = (DriveFsmStorage *)context;
+  return e->data == storage->destination;
+}
+
+FSM_STATE_TRANSITION(state_set_motorcontroller_output) {
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_DRIVE,
+                             prv_destination_guard, state_set_relays);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_REVERSE,
+                             prv_destination_guard, state_set_relays);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_SET_OUTPUT_DESTINATION_NEUTRAL_PARKING,
+                             prv_destination_guard, state_neutral);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+FSM_STATE_TRANSITION(state_set_relays) {
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_CLOSED_DESTINATION_DRIVE,
+                             prv_destination_guard, state_drive);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_CLOSED_DESTINATION_REVERSE,
+                             prv_destination_guard, state_reverse);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_RELAYS_OPENED_DESTINATION_NEUTRAL_PARKING,
+                             prv_destination_guard, state_set_motorcontroller_output);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_fault);
+}
+
+FSM_STATE_TRANSITION(state_set_ebrake) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_EBRAKE_PRESSED, state_parking);
+  FSM_ADD_GUARDED_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_EBRAKE_RELEASED, prv_destination_guard,
+                             state_neutral);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_ebrake_fault);
+}
+
+FSM_STATE_TRANSITION(state_ebrake_fault) {
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_FAULT, state_ebrake_fault);
+  FSM_ADD_TRANSITION(DRIVE_FSM_INPUT_EVENT_MCI_EBRAKE_RELEASED, state_neutral);
+}
+
+typedef struct EventPair {
+  EventId success_event;
+  EventId fault_event;
+} EventPair;
```

Comment:
whoa define struct in c file :eyes:

Reply:
it's only used internally :P rest of the modules don't need to know about this.